docs/en/Community-Articles/2024-05-09-what-is-object-to-object-mapping/post.md

```diff
@@ -0,0 +1,122 @@
+        );
+        CustomerDto customerDto = new CustomerDto
+        {
+            Id = customer.Id,
+            FirstName = input.FirstName,
+            LastName = input.LastName,
+            // ...other
+        };
+        
+        return customerDto;
+    }
+
+````
+This code is repetitive, tedious to write. It violates DRY principles, makes your code very complex and reduces readability.  Instead of using this approach, you can use the automapper which does the matching automatically. Here is a better code:
+
+
+````csharp
+        public virtual async Task<CustomerDto> CreateAsync(CustomerCreateDto input)
+        {
+
+            var customer = await _customerManager.CreateAsync(
+            input.BirthDay, input.MembershipDate, input.FirstName, input.LastName
+            );
+
+            return ObjectMapper.Map<Customer, CustomerDto>(customer);
+        }
+````
+
+# What is AutoMapper?
+
+![Swagger](./images/automapper.png)
+
+Automapper is a .net library that automates object to object mapping. The implementation of Automapper is quite simple. In this chapter I will show you how to use the Automapper library in an ABP-based application. For this reason, I assume that you already have an ABP-based application created. If you have not yet created an ABP-based application, please follow the [Getting Started documentation](https://docs.abp.io/en/abp/latest/Getting-Started-Create-Solution?UI=MVC&DB=EF&Tiered=No).
+
+Create a domain entity similar to this one: 
+
+````csharp
+    public class Customer : FullAuditedAggregateRoot<Guid>
+    {
+        public string? FirstName { get; set; }
+        public string? LastName { get; set; }
+        public DateTime BirthDay { get; set; }
+        public DateTime MembershipDate { get; set; }
+    }
+````
+`Custumer` contains a lot of attributes. we don't want to give all of these attributes to users, so create dto with only the necessary attributes: 
+
+````csharp
+    public class CustomerGetDto
+    {
+        public string? FirstName { get; set; }
+        public string? LastName { get; set; }
+        public DateTime BirthDay { get; set; }
+    }
+````
+
+You can use this code to list customers:
+
+````csharp
+public virtual async Task<PagedResultDto<CustomerGetDto>> GetListAsync(GetCustomersInput input)
+ {
+     var totalCount = await _customerRepository.GetCountAsync(input.FilterText, input.FirstName, input.LastName, input.BirthDayMin, input.BirthDayMax, input.MembershipDateMin, input.MembershipDateMax);
+     var items = await _customerRepository.GetListAsync(input.FilterText, input.FirstName, input.LastName, input.BirthDayMin, input.BirthDayMax, input.MembershipDateMin, input.MembershipDateMax, input.Sorting, input.MaxResultCount, input.SkipCount);
+
+     return new PagedResultDto<CustomerGetDto>
+     {
+         TotalCount = totalCount,
+         Items = ObjectMapper.Map<List<Customer>, List<CustomerGetDto>>(items)
+     };
+ }
+````
+
```

Comment:
You should write a small explanation before/after each code block in articles. You can write what this code block does, for example:

> Here, we basically get the total count of our customers and all of the customers according to the filters specified and then by using the `ObjectMapper.Map` method, which came from the [`ApplicationService` base class](https://docs.abp.io/en/abp/latest/Application-Services), we are mapping the `List<Customer>` to `List<CustomerGetDto>`. Thanks to that, we have full control over which properties to be returned to the end-users.

Reply:
I added this sentence:
In this code, we first get the total number of our customers and all customers according to the specified filters, then we map List<Customer> to List<CustomerGetDto> using the ObjectMapper.Map method from the ApplicationService base class. This way we have full control over which properties are returned to the end users.